File: src/Recetas.Web/Servicios/ClienteDeApi.cs
using System.Net;
using System.Net.Http.Json;
using Recetas.Contratos.Contrasenas;
using Recetas.Contratos.Cuentas;
using Recetas.Contratos.Moderacion;
using Recetas.Contratos.Recetas;
using Recetas.Contratos.Registro;
using Recetas.Contratos.Sesiones;

namespace Recetas.Web.Servicios;

/// <summary>
/// Envoltorio de las llamadas a la API. Traduce respuestas HTTP a resultados que
/// las páginas puedan mostrar, sin que cada una repita el manejo de errores.
/// </summary>
public sealed class ClienteDeApi(HttpClient http)
{
    private const string MensajeDeRedCaida =
        "No se ha podido contactar con el servidor. Comprueba tu conexión.";

    // ------------------------------------------------------------- Cuentas

    public Task<ResultadoDeLlamada> SolicitarAltaAsync(string correo) =>
        EnviarAsync(HttpMethod.Post, "registro/solicitudes",
            new PeticionDeSolicitudDeRegistro { Correo = correo });

    public Task<ResultadoDeLlamada> CompletarAltaAsync(string token, string contrasena) =>
        EnviarAsync(HttpMethod.Post, "registro/completar",
            new PeticionDeCompletarRegistro { Token = token, Contrasena = contrasena });

    public Task<ResultadoDeLlamada> SolicitarContrasenaAsync(string correo) =>
        EnviarAsync(HttpMethod.Post, "contrasena/solicitudes",
            new PeticionDeSolicitudDeContrasena { Correo = correo });

    public Task<ResultadoDeLlamada> RestablecerContrasenaAsync(string token, string contrasena) =>
        EnviarAsync(HttpMethod.Post, "contrasena/restablecer",
            new PeticionDeRestablecerContrasena { Token = token, Contrasena = contrasena });

    /// <summary>Cambia la contraseña con la sesión iniciada, sabiendo la actual.</summary>
    public Task<ResultadoDeLlamada> CambiarContrasenaAsync(string actual, string nueva) =>
        EnviarAsync(HttpMethod.Put, "yo/contrasena",
            new PeticionDeCambioDeContrasena { ContrasenaActual = actual, ContrasenaNueva = nueva });

    public async Task<(ResultadoDeLlamada Resultado, RespuestaDeInicioDeSesion? Acceso)> IniciarSesionAsync(
        string correo,
        string contrasena)
    {
        try
        {
            var respuesta = await http.PostAsJsonAsync("sesiones",
                new PeticionDeInicioDeSesion { Correo = correo, Contrasena = contrasena });

            if (!respuesta.IsSuccessStatusCode)
            {
                return (await TraducirErrorAsync(respuesta), null);
            }

            return (ResultadoDeLlamada.Correcto(),
                await respuesta.Content.ReadFromJsonAsync<RespuestaDeInicioDeSesion>());
        }
        catch (HttpRequestException)
        {
            return (ResultadoDeLlamada.Fallido(MensajeDeRedCaida), null);
        }
    }

    // ------------------------------------------------------------- Recetas

    public Task<(ResultadoDeLlamada Resultado, List<ResumenDeReceta>? Recetas)> ListarRecetasAsync() =>
        LeerAsync<List<ResumenDeReceta>>("recetas");

    /// <param name="raciones">
    /// Comensales a los que ajustar las cantidades. El servidor hace el cálculo y el
    /// redondeo: son reglas de negocio y no se duplican aquí.
    /// </param>
    public Task<(ResultadoDeLlamada Resultado, RespuestaDeReceta? Receta)> ObtenerRecetaAsync(
        Guid id,
        int? raciones = null) =>
        LeerAsync<RespuestaDeReceta>(
            raciones is { } valor ? $"recetas/{id}?raciones={valor}" : $"recetas/{id}");

    public async Task<(ResultadoDeLlamada Resultado, Guid Id)> CrearRecetaAsync(PeticionDeReceta peticion)
    {
        try
        {
            var respuesta = await http.PostAsJsonAsync("recetas", peticion);

            if (!respuesta.IsSuccessStatusCode)
            {
                return (await TraducirErrorAsync(respuesta), Guid.Empty);
            }

            var creada = await respuesta.Content.ReadFromJsonAsync<RespuestaDeReceta>();
            return (ResultadoDeLlamada.Correcto(), creada?.Id ?? Guid.Empty);
        }
        catch (HttpRequestException)
        {
            return (ResultadoDeLlamada.Fallido(MensajeDeRedCaida), Guid.Empty);
        }
    }

    /// <summary>
    /// Pide al servidor que lea una receta de una página web. Devuelve un borrador:
    /// no crea nada hasta que el usuario envía el formulario.
    /// </summary>
    public async Task<(ResultadoDeLlamada Resultado, RespuestaDeImportacion? Borrador)> ImportarRecetaAsync(
        string direccion)
    {
        try
        {
            var respuesta = await http.PostAsJsonAsync("recetas/importaciones",
                new PeticionDeImportacion { Direccion = direccion });

            if (!respuesta.IsSuccessStatusCode)
            {
                return (await TraducirErrorAsync(respuesta), null);
            }

            return (ResultadoDeLlamada.Correcto(),
                await respuesta.Content.ReadFromJsonAsync<RespuestaDeImportacion>());
        }
        catch (HttpRequestException)
        {
            return (ResultadoDeLlamada.Fallido(MensajeDeRedCaida), null);
        }
    }

    public Task<ResultadoDeLlamada> ActualizarRecetaAsync(Guid id, PeticionDeReceta peticion) =>
        EnviarAsync(HttpMethod.Put, $"recetas/{id}", peticion);

    public Task<ResultadoDeLlamada> BorrarRecetaAsync(Guid id) =>
        EnviarAsync(HttpMethod.Delete, $"recetas/{id}");

    // --------------------------------------------------------- Publicación

    public Task<ResultadoDeLlamada> PublicarAsync(Guid id) =>
        EnviarAsync(HttpMethod.Post, $"recetas/{id}/publicacion");

    public Task<ResultadoDeLlamada> DespublicarAsync(Guid id) =>
        EnviarAsync(HttpMethod.Delete, $"recetas/{id}/publicacion");

    // ----------------------------------------------------------- Favoritos

    /// <summary>
    /// Mis favoritas <b>que todavía puedo ver</b>. Lo segundo lo decide el
    /// servidor: una receta marcada y luego despublicada no vuelve en esta lista.
    /// </summary>
    public Task<(ResultadoDeLlamada Resultado, List<ResumenDeReceta>? Recetas)> ListarFavoritasAsync() =>
        LeerAsync<List<ResumenDeReceta>>("recetas/favoritas");

    public Task<ResultadoDeLlamada> MarcarFavoritaAsync(Guid id) =>
        EnviarAsync(HttpMethod.Put, $"recetas/{id}/favorito");

    public Task<ResultadoDeLlamada> DesmarcarFavoritaAsync(Guid id) =>
        EnviarAsync(HttpMethod.Delete, $"recetas/{id}/favorito");

    // -------------------------------------------------------------- Cuenta

    public Task<(ResultadoDeLlamada Resultado, RespuestaDeIdentidad? Identidad)> ObtenerIdentidadAsync() =>
        LeerAsync<RespuestaDeIdentidad>("yo");

    /// <summary>
    /// Descarga el paquete con los datos del usuario.
    /// </summary>
    /// <remarks>
    /// Lo pide este cliente y no un enlace del navegador porque una descarga
    /// iniciada por el navegador no lleva la cabecera de autorización, y meter el
    /// testigo en la dirección lo dejaría escrito en los registros del servidor y
    /// en el historial.
    /// </remarks>
    public async Task<(ResultadoDeLlamada Resultado, Stream? Contenido)> DescargarMisDatosAsync()
    {
        try
        {
            // Sin esperar al cuerpo entero: el paquete puede pesar y no hay razón
            // para tenerlo dos veces en memoria.
            var respuesta = await http.GetAsync(
                "yo/datos",
                HttpCompletionOption.ResponseHeadersRead);

            if (!respuesta.IsSuccessStatusCode)
            {
                return (await TraducirErrorAsync(respuesta), null);
            }

            return (ResultadoDeLlamada.Correcto(), await respuesta.Content.ReadAsStreamAsync());
        }
        catch (HttpRequestException)
        {
            return (ResultadoDeLlamada.Fallido(MensajeDeRedCaida), null);
        }
    }

    /// <summary>
    /// Borra la cuenta. La contraseña va en el cuerpo de un DELETE: en la URL
    /// acabaría en los registros del servidor y en el historial del navegador.
    /// </summary>
    public Task<ResultadoDeLlamada> BorrarMiCuentaAsync(string contrasena) =>
        EnviarAsync(HttpMethod.Delete, "yo", new PeticionDeBaja { Contrasena = contrasena });

    // ----------------------------------------------------------- Denuncias

    public Task<ResultadoDeLlamada> DenunciarAsync(Guid recetaId, string motivo, string? comentario) =>
        EnviarAsync(
            HttpMethod.Post,
            $"recetas/{recetaId}/denuncias",
            new PeticionDeDenuncia { Motivo = motivo, Comentario = comentario });

    // --------------------------------------------------------------- Fotos

    public async Task<ResultadoDeLlamada> SubirFotoAsync(Guid recetaId, Stream contenido)
    {
        try
        {
            using var cuerpo = new StreamContent(contenido);
            var respuesta = await http.PostAsync($"recetas/{recetaId}/fotos", cuerpo);

            return respuesta.IsSuccessStatusCode
                ? ResultadoDeLlamada.Correcto()
                : await TraducirErrorAsync(respuesta);
        }
        catch (HttpRequestException)
        {
            return ResultadoDeLlamada.Fallido(MensajeDeRedCaida);
        }
    }

    /// <summary>
    /// Descarga una foto y la devuelve como URL <c>data:</c>, lista para el
    /// atributo <c>src</c> de una imagen.
    /// </summary>
    /// <remarks>
    /// Una etiqueta <c>&lt;img&gt;</c> no manda la cabecera <c>Authorization</c>,
    /// y el endpoint la exige, así que la imagen hay que traerla con este cliente.
    /// <para>
    /// Se devuelve como <c>data:</c> y no como URL de objeto (<c>blob:</c>) porque
    /// la política de seguridad de contenido del sitio permite la primera y no la
    /// segunda; cambiarla exigiría tocar nginx en el servidor. El precio es el
    /// 33 % que engorda base64.
    /// </para>
    /// </remarks>
    public Task<string?> DescargarFotoAsync(Guid recetaId, Guid fotoId) =>
        DescargarImagenAsync($"recetas/{recetaId}/fotos/{fotoId}");

    /// <summary>
    /// Igual, pero la versión reducida. Es la que usan los listados: pintar
    /// cincuenta tarjetas con el archivo original serían cientos de megabytes,
    /// y encima inflados un 33 % por el base64.
    /// </summary>
    public Task<string?> DescargarMiniaturaAsync(Guid recetaId, Guid fotoId) =>
        DescargarImagenAsync($"recetas/{recetaId}/fotos/{fotoId}/miniatura");

    private async Task<string?> DescargarImagenAsync(string ruta)
    {
        try
        {
            var respuesta = await http.GetAsync(ruta);

            if (!respuesta.IsSuccessStatusCode)
            {
                return null;
            }

            var bytes = await respuesta.Content.ReadAsByteArrayAsync();
            var tipo = respuesta.Content.Headers.ContentType?.MediaType ?? "image/jpeg";

            return $"data:{tipo};base64,{Convert.ToBase64String(bytes)}";
        }
        catch (HttpRequestException)
        {
            return null;
        }
    }

    public Task<ResultadoDeLlamada> BorrarFotoAsync(Guid recetaId, Guid fotoId) =>
        EnviarAsync(HttpMethod.Delete, $"recetas/{recetaId}/fotos/{fotoId}");

    // ------------------------------------------------------------ Búsqueda

    public Task<(ResultadoDeLlamada Resultado, RespuestaDeBusqueda? Busqueda)> BuscarAsync(
        string? nombre,
        IEnumerable<string> ingredientes,
        string? tipo,
        IEnumerable<string>? etiquetas = null)
    {
        var partes = new List<string>();

        if (!string.IsNullOrWhiteSpace(nombre))
        {
            partes.Add($"nombre={Uri.EscapeDataString(nombre)}");
        }

        foreach (var ingrediente in ingredientes.Where(valor => !string.IsNullOrWhiteSpace(valor)))
        {
            partes.Add($"ingrediente={Uri.EscapeDataString(ingrediente)}");
        }

        foreach (var etiqueta in (etiquetas ?? []).Where(valor => !string.IsNullOrWhiteSpace(valor)))
        {
            partes.Add($"etiqueta={Uri.EscapeDataString(etiqueta)}");
        }

        if (!string.IsNullOrWhiteSpace(tipo))
        {
            partes.Add($"tipo={Uri.EscapeDataString(tipo)}");
        }

        var consulta = partes.Count > 0 ? "?" + string.Join("&", partes) : string.Empty;

        return LeerAsync<RespuestaDeBusqueda>($"recetas/busqueda{consulta}");
    }

    // ----------------------------------------------------------- Utilidades

    private async Task<(ResultadoDeLlamada Resultado, T? Datos)> LeerAsync<T>(string ruta)
    {
        try
        {
            var respuesta = await http.GetAsync(ruta);

            if (!respuesta.IsSuccessStatusCode)
            {
                return (await TraducirErrorAsync(respuesta), default);
            }

            return (ResultadoDeLlamada.Correcto(), await respuesta.Content.ReadFromJsonAsync<T>());
        }
        catch (HttpRequestException)
        {
            return (ResultadoDeLlamada.Fallido(MensajeDeRedCaida), default);
        }
    }

    /// <summary>Petición sin cuerpo: borrados y cambios de estado.</summary>
    private Task<ResultadoDeLlamada> EnviarAsync(HttpMethod metodo, string ruta) =>
        EnviarAsync<object>(metodo, ruta, null);

    private async Task<ResultadoDeLlamada> EnviarAsync<TPeticion>(
        HttpMethod metodo,
        string ruta,
        TPeticion? contenido)
    {
        try
        {
            using var peticion = new HttpRequestMessage(metodo, ruta);

            if (contenido is not null)
            {
                peticion.Content = JsonContent.Create(contenido);
            }

            var respuesta = await http.SendAsync(peticion);

            if (!respuesta.IsSuccessStatusCode)
            {
                return await TraducirErrorAsync(respuesta);
            }

            // Las respuestas 204 no traen cuerpo; las 200 de esta API traen un
            // mensaje que conviene mostrar.
            if (respuesta.StatusCode == HttpStatusCode.NoContent)
            {
                return ResultadoDeLlamada.Correcto();
            }

            var cuerpo = await respuesta.Content.ReadFromJsonAsync<RespuestaDeError>();
            return ResultadoDeLlamada.Correcto(cuerpo?.Mensaje);
        }
        catch (HttpRequestException)
        {
            return ResultadoDeLlamada.Fallido(MensajeDeRedCaida);
        }
        catch (NotSupportedException)
        {
            // Cuerpo que no es JSON en un error inesperado.
            return ResultadoDeLlamada.Fallido("Respuesta inesperada del servidor.");
        }
    }

    private static async Task<ResultadoDeLlamada> TraducirErrorAsync(HttpResponseMessage respuesta)
    {
        if (respuesta.StatusCode == HttpStatusCode.TooManyRequests)
        {
            return ResultadoDeLlamada.Fallido("Demasiados intentos. Espera unos minutos y vuelve a probar.");
        }

        if (respuesta.StatusCode == HttpStatusCode.RequestEntityTooLarge)
        {
            return ResultadoDeLlamada.Fallido("El archivo es demasiado grande.");
        }

        try
        {
            var cuerpo = await respuesta.Content.ReadFromJsonAsync<RespuestaDeError>();

            if (!string.IsNullOrWhiteSpace(cuerpo?.Mensaje))
            {
                return ResultadoDeLlamada.Fallido(cuerpo.Mensaje);
            }
        }
        catch (Exception)
        {
            // La validación automática de ASP.NET responde en otro formato.
        }

        return respuesta.StatusCode switch
        {
            HttpStatusCode.NotFound => ResultadoDeLlamada.Fallido("No se ha encontrado."),
            HttpStatusCode.BadRequest => ResultadoDeLlamada.Fallido("Revisa los datos introducidos."),
            _ => ResultadoDeLlamada.Fallido("No se ha podido completar la operación. Inténtalo de nuevo.")
        };
    }
}

/// <param name="Correcta">Si la operación salió bien.</param>
/// <param name="Mensaje">Texto a mostrar al usuario, ya sea de éxito o de error.</param>
public sealed record ResultadoDeLlamada(bool Correcta, string? Mensaje)
{
    public static ResultadoDeLlamada Correcto(string? mensaje = null) => new(true, mensaje);

    public static ResultadoDeLlamada Fallido(string mensaje) => new(false, mensaje);
}

File: android/app/src/main/java/com/angelgf/recetas/datos/ClienteDeApi.kt
package com.angelgf.recetas.datos

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.defaultRequest
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.readRawBytes
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

/**
 * Resultado de una llamada. Se prefiere esto a lanzar excepciones porque cada
 * pantalla tiene que pintar el error, no tragárselo.
 */
sealed interface Resultado<out T> {
    data class Correcto<T>(val valor: T) : Resultado<T>
    data class Fallo(val mensaje: String) : Resultado<Nothing>

    /** El token ya no vale. Lo trata la aplicación entera igual: volver al login. */
    data object SesionCaducada : Resultado<Nothing>
}

/**
 * Cliente de la API.
 *
 * Dos cosas viven aquí y en ningún otro sitio, igual que en
 * `ManejadorDeAutenticacion` de la web y por el mismo motivo —que ninguna pantalla
 * pueda olvidarse—:
 *
 * - La cabecera `Authorization`.
 * - El `401`, que se convierte en [Resultado.SesionCaducada].
 */
class ClienteDeApi(
    private val baseDeLaApi: String,
    private val sesion: AlmacenDeSesion,
    motor: HttpClientEngine? = null
) {
    private val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
    }

    private val http: HttpClient = crearCliente(motor)

    private fun crearCliente(motor: HttpClientEngine?): HttpClient {
        val configuracion: io.ktor.client.HttpClientConfig<*>.() -> Unit = {
            install(ContentNegotiation) { json(this@ClienteDeApi.json) }

            defaultRequest {
                url(baseDeLaApi)
                sesion.leer()?.let { header("Authorization", "Bearer $it") }
            }
        }

        return if (motor is HttpClientEngine) HttpClient(motor, configuracion) else HttpClient(configuracion)
    }

    // ------------------------------------------------------------ Sesión

    suspend fun iniciarSesion(correo: String, contrasena: String): Resultado<Unit> = envolver {
        val respuesta = http.post("sesiones") {
            contentType(ContentType.Application.Json)
            setBody(PeticionDeInicioDeSesion(correo, contrasena))
        }

        if (!respuesta.status.isSuccess()) {
            // El inicio de sesión es el único sitio donde un 401 NO significa
            // "sesión caducada": significa que las credenciales no valen.
            return@envolver Resultado.Fallo(
                if (respuesta.status == HttpStatusCode.Unauthorized) {
                    "Correo o contraseña incorrectos."
                } else {
                    mensajeDeError(respuesta)
                }
            )
        }

        sesion.guardar(respuesta.body<RespuestaDeInicioDeSesion>().token)
        Resultado.Correcto(Unit)
    }

    fun cerrarSesion() = sesion.borrar()

    fun haySesion(): Boolean = sesion.leer() != null

    suspend fun yo(): Resultado<RespuestaDeIdentidad> = leer("yo")

    // ------------------------------------------------------------ Recetas

    suspend fun misRecetas(): Resultado<List<ResumenDeReceta>> = leer("recetas")

    /**
     * @param raciones comensales a los que ajustar las cantidades. El cálculo y el
     *   redondeo los hace el servidor: son reglas de negocio (feature 010) y no se
     *   reimplementan aquí.
     */
    suspend fun receta(id: String, raciones: Int? = null): Resultado<RespuestaDeReceta> =
        envolver {
            val respuesta = http.get("recetas/$id") {
                raciones?.let { parameter("raciones", it) }
            }

            traducir(respuesta)
        }

    suspend fun crearReceta(peticion: PeticionDeReceta): Resultado<RespuestaDeReceta> =
        envolver {
            val respuesta = http.post("recetas") {
                contentType(ContentType.Application.Json)
                setBody(peticion)
            }

            traducir(respuesta)
        }

    suspend fun actualizarReceta(id: String, peticion: PeticionDeReceta): Resultado<Unit> =
        envolver {
            val respuesta = http.put("recetas/$id") {
                contentType(ContentType.Application.Json)
                setBody(peticion)
            }

            sinContenido(respuesta)
        }

    suspend fun borrarReceta(id: String): Resultado<Unit> =
        envolver { sinContenido(http.delete("recetas/$id")) }

    // ------------------------------------------------------- Publicación

    suspend fun publicar(id: String): Resultado<Unit> =
        envolver { sinContenido(http.post("recetas/$id/publicacion")) }

    suspend fun despublicar(id: String): Resultado<Unit> =
        envolver { sinContenido(http.delete("recetas/$id/publicacion")) }

    // -------------------------------------------------------- Denuncias

    suspend fun denunciar(recetaId: String, motivo: String, comentario: String?): Resultado<Unit> =
        envolver {
            val respuesta = http.post("recetas/$recetaId/denuncias") {
                contentType(ContentType.Application.Json)
                setBody(PeticionDeDenuncia(motivo, comentario?.takeIf { it.isNotBlank() }))
            }

            sinContenido(respuesta)
        }

    // ------------------------------------------------------------- Fotos

    suspend fun subirFoto(recetaId: String, contenido: ByteArray): Resultado<Unit> =
        envolver {
            val respuesta = http.post("recetas/$recetaId/fotos") {
                setBody(contenido)
            }

            traducirVacio(respuesta)
        }

    suspend fun borrarFoto(recetaId: String, fotoId: String): Resultado<Unit> =
        envolver { sinContenido(http.delete("recetas/$recetaId/fotos/$fotoId")) }

    // --------------------------------------------------------- Importar

    suspend fun importar(direccion: String): Resultado<RespuestaDeImportacion> =
        envolver {
            val respuesta = http.post("recetas/importaciones") {
                contentType(ContentType.Application.Json)
                setBody(PeticionDeImportacion(direccion))
            }

            traducir(respuesta)
        }

    // ------------------------------------------------------------ Cuenta

    suspend fun solicitarAlta(correo: String): Resultado<String> =
        mensajeDe("registro/solicitudes", PeticionDeSolicitudDeRegistro(correo))

    suspend fun completarAlta(token: String, contrasena: String): Resultado<String> =
        mensajeDe("registro/completar", PeticionDeCompletarRegistro(token, contrasena))

    suspend fun solicitarContrasena(correo: String): Resultado<String> =
        mensajeDe("contrasena/solicitudes", PeticionDeSolicitudDeContrasena(correo))

    suspend fun restablecerContrasena(token: String, contrasena: String): Resultado<String> =
        mensajeDe("contrasena/restablecer", PeticionDeRestablecerContrasena(token, contrasena))

    /** Endpoints que responden 200 con un mensaje pensado para enseñárselo al usuario. */
    private suspend inline fun <reified T> mensajeDe(ruta: String, cuerpo: T): Resultado<String> =
        envolver {
            val respuesta = http.post(ruta) {
                contentType(ContentType.Application.Json)
                setBody(cuerpo)
            }

            if (respuesta.status.isSuccess()) {
                Resultado.Correcto(respuesta.body<RespuestaDeError>().mensaje.orEmpty())
            } else {
                Resultado.Fallo(mensajeDeError(respuesta))
            }
        }

    suspend fun buscar(nombre: String, ingredientes: List<String>): Resultado<RespuestaDeBusqueda> =
        envolver {
            val respuesta = http.get("recetas/busqueda") {
                if (nombre.isNotBlank()) parameter("nombre", nombre)

                // El parámetro se repite, no se separa por comas: es como lo
                // espera la API desde la feature 006.
                ingredientes.filter { it.isNotBlank() }.forEach { parameter("ingrediente", it) }
            }

            traducir(respuesta)
        }

    /**
     * Descarga una imagen como bytes.
     *
     * No sirve una biblioteca de carga de imágenes apuntada a la URL: el endpoint
     * de fotos exige la cabecera de autorización, y una etiqueta de imagen normal
     * no la manda. Es el mismo problema que la web resuelve con URLs `data:`.
     */
    suspend fun miniatura(recetaId: String, fotoId: String): ByteArray? =
        imagen("recetas/$recetaId/fotos/$fotoId/miniatura")

    suspend fun foto(recetaId: String, fotoId: String): ByteArray? =
        imagen("recetas/$recetaId/fotos/$fotoId")

    private suspend fun imagen(ruta: String): ByteArray? =
        try {
            val respuesta = http.get(ruta)
            if (respuesta.status.isSuccess()) respuesta.readRawBytes() else null
        } catch (excepcion: Exception) {
            // Una foto que no carga deja un hueco; no debe tumbar la pantalla.
            null
        }

    // --------------------------------------------------------- Utilidades

    private suspend inline fun <reified T> leer(ruta: String): Resultado<T> =
        envolver { traducir(http.get(ruta)) }

    /** Para respuestas 204, que no traen cuerpo que interpretar. */
    private suspend fun sinContenido(respuesta: HttpResponse): Resultado<Unit> =
        traducirVacio(respuesta)

    private suspend fun traducirVacio(respuesta: HttpResponse): Resultado<Unit> =
        when {
            respuesta.status.isSuccess() -> Resultado.Correcto(Unit)
            respuesta.status == HttpStatusCode.Unauthorized -> {
                sesion.borrar()
                Resultado.SesionCaducada
            }
            else -> Resultado.Fallo(mensajeDeError(respuesta))
        }

    private suspend inline fun <reified T> traducir(respuesta: HttpResponse): Resultado<T> =
        when {
            respuesta.status.isSuccess() -> Resultado.Correcto(respuesta.body<T>())
            respuesta.status == HttpStatusCode.Unauthorized -> {
                // El token ya no vale: se borra aquí para que no quede una sesión
                // muerta en el dispositivo esperando a fallar otra vez.
                sesion.borrar()
                Resultado.SesionCaducada
            }
            else -> Resultado.Fallo(mensajeDeError(respuesta))
        }

    /**
     * Traduce el estado HTTP a algo que un usuario entienda.
     *
     * Se intenta primero el mensaje que manda la API, que está escrito para leerse;
     * si no viene, se cae a un texto por código.
     */
    suspend fun mensajeDeError(respuesta: HttpResponse): String {
        val delServidor = try {
            respuesta.body<RespuestaDeError>().mensaje
        } catch (excepcion: Exception) {
            null
        }

        if (!delServidor.isNullOrBlank()) return delServidor

        return when (respuesta.status) {
            HttpStatusCode.NotFound -> "No se ha encontrado."
            HttpStatusCode.TooManyRequests -> "Demasiados intentos. Espera unos minutos."
            HttpStatusCode.BadRequest -> "Revisa los datos introducidos."
            else -> "No se ha podido completar la operación. Inténtalo de nuevo."
        }
    }

    private suspend fun <T> envolver(bloque: suspend () -> Resultado<T>): Resultado<T> =
        try {
            bloque()
        } catch (excepcion: Exception) {
            // Sin red, DNS caído, servidor apagado. El usuario necesita saber que
            // el problema es la conexión y no sus datos.
            Resultado.Fallo("No se ha podido contactar con el servidor. Comprueba tu conexión.")
        }
}

private fun HttpStatusCode.isSuccess(): Boolean = value in 200..299
